src/connector/src/sink/snowflake.rs

```diff
@@ -0,0 +1,279 @@
+            config.common.user.clone(),
+            config.common.database.clone(),
+            config.common.schema.clone(),
+            config.common.pipe.clone(),
+            config.common.rsa_public_key_fp.clone(),
+            config.common.private_key.clone(),
+            HashMap::new(),
+        );
+
+        let s3_client = SnowflakeS3Client::new(
+            config.common.s3_bucket.clone(),
+            config.common.aws_access_key_id.clone(),
+            config.common.aws_secret_access_key.clone(),
+            config.common.aws_region.clone(),
+        )
+        .await;
+
+        Self {
+            config,
+            schema: schema.clone(),
+            pk_indices,
+            is_append_only,
+            http_client,
+            s3_client,
+            row_encoder: JsonEncoder::new(
+                schema,
+                None,
+                super::encoder::DateHandlingMode::String,
+                TimestampHandlingMode::String,
+                TimestamptzHandlingMode::UtcString,
+                TimeHandlingMode::String,
+            ),
+            row_counter: 0,
+            payload: String::new(),
+            // Start from 0, i.e., `RW_SNOWFLAKE_S3_SINK_FILE_0`
+            sink_file_suffix: 0,
+        }
+    }
+
+    /// Note that we shall NOT reset the `sink_file_suffix`
+    /// since we need to incrementally keep the sink
+    /// file *unique*, otherwise snowflake will not
+    /// sink it from external stage (i.e., s3)
+    fn reset(&mut self) {
+        self.payload.clear();
+        self.row_counter = 0;
+    }
+
+    fn at_sink_threshold(&self) -> bool {
+        self.row_counter >= MAX_BATCH_ROW_NUM
+    }
+
+    fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op != Op::Insert {
+                continue;
+            }
+            let row_json_string = Value::Object(self.row_encoder.encode(row)?).to_string();
+            self.payload.push_str(&row_json_string);
+            self.row_counter += 1;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SinkWriter for SnowflakeSinkWriter {
+    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn update_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) -> Result<()> {
+        Ok(())
+    }
+
+    async fn barrier(&mut self, _is_checkpoint: bool) -> Result<Self::CommitMetadata> {
+        Ok(())
+    }
+
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        self.append_only(chunk)?;
+
+        // When the number of row exceeds `MAX_BATCH_ROW_NUM`
+        if self.at_sink_threshold() {
+            // first sink to the external stage provided by user (i.e., s3)
+            self.s3_client
+                .sink_to_s3(self.payload.clone().into(), self.sink_file_suffix)
+                .await?;
```

Comment:
Will it block the streaming when the S3 write is slow? If so, I recommend to add an `await_tree::Span` to it.

Reply:
I think so, since `sink_to_s3` is essentially a blocking call.

Reply:
>  If so, I recommend to add an await_tree::Span to it.

do you mean the combination use of `instrument_await` + `await_tree::spawn`?

ps. I personally prefer to postpone this to subsequent pr(s).